Changelogs of v1.7.0 (#61)
meilisearch/engine-team: meilisearch-release-changelog/v1.7.0.md

  },
  "small": {
    "source": "openAi",
    "model": "text-embedding-3-small",
    "dimensions": 1024 // must be >0, must be <= 1536 for "text-embedding-3-small"
  },
  "legacy": {
    "source": "openAi",
    "model": "text-embedding-ada-002",
    "dimensions": 1536 // must =1536  for "text-embedding-ada-002"
  },
  "omitted_dimensions": { // uses the default dimension
    "source": "openAi",
    "model": "text-embedding-ada-002",
  }
}
```

Done in #4375 by Gosti.

### Add GPU support to compute embeddings

Enabling the CUDA feature allows using an available GPU to compute embeddings with a `huggingFace` embedder.
On an AWS Graviton 2, this yields a x3 - x5 improvement on indexing time.

👇 How to enable GPU support through CUDA for HuggingFace embedding generation:

**Prerequisites**

- Linux distribution with a compatible CUDA version
- NVidia GPU with CUDA support
- A recent Rust compiler to compile Meilisearch from source

**Steps**

1. Follow [the guide](https://huggingface.github.io/candle/guide/installation.html) to install the CUDA dependencies
2. Clone Meilisearch: `git clone https://github.com/meilisearch/meilisearch.git`

> 3. Compile Meilisearch with the `cuda` feature: `cargo build --release --features cuda`

Reviewer: By adding the `--package` flag you force cargo to only compile the meilisearch binary, nothing more.

```suggestion
3. Compile Meilisearch with the `cuda` feature: `cargo build --release --package meilisearch --features cuda`
```

  - `target`: selects what logs you’re interested in. It takes the form of `code_part=log_level`. For example, `index_scheduler=info`
  - `mode`: selects in what format of log you want. Two options are available: `fmt` (basic logs) or `profile` (a way more complex trace)
- `DELETE /logs/stream`: stops the listener from the meilisearch perspective. Does not require any parameters.

💡 *More information in the [New experimental routes section of this file](https://meilisearch.notion.site/Logs-trace-changes-in-v1-7-0-6dd111ecd66448aba30d2f3cef212749).*

⚠️ Some remarks on this `POST /logs/stream` route:
- You can have only one listener at a time
- Listening to the route doesn’t seem to work with `xh` or `httpie` for the moment
- When killing the listener, it may stay installed on Meilisearch for some time, and you will need to call the `DELETE /logs/stream` route to get rid of it.

🗣️ This feature is experimental, and we need your help to improve it! Share your thoughts and feedback on this [GitHub discussion](https://github.com/orgs/meilisearch/discussions/721).

⚠️ Experimental features may be incompatible between Meilisearch versions.

## Other improvements

> * Related to the [Prometheus experimental feauture](https://github.com/orgs/meilisearch/discussions/625): add job variable to Grafana dashboard (#4330) @capJavert

Reviewer: ```suggestion
* Related to the [Prometheus experimental feature](https://github.com/orgs/meilisearch/discussions/625): add job variable to Grafana dashboard (#4330) @capJavert
```

```

To this:

```bash
2024-02-06T13:58:14.710803Z  INFO actix_server::builder: 200: starting 10 workers
```

### Experimental: choose your log format

And one new CLI parameter is introduced called `--experimental_logs_format`. This lets you specify the log format you want to use on the console.

Accepted values are:

- `human`: the default human-readable one
- `json`: JSON structured logs

🗣️ This feature is experimental, and we need your help to improve it! Share your thoughts and feedback on this [GitHub discussion](TBD).

⚠️ Experimental features may be incompatible between Meilisearch versions.

### Experimental: new routes to manage logs

This new version of Meilisearch introduces 3 new experimental routes
- `POST /logs/stream`: streams the log happening in real-time. Requires two parameters:
  - `target`: selects what logs you’re interested in. It takes the form of `code_part=log_level`. For example, `index_scheduler=info`
  - `mode`: selects in what format of log you want. Two options are available: `fmt` (basic logs) or `profile` (a way more complex trace)
- `DELETE /logs/stream`: stops the listener from the meilisearch perspective. Does not require any parameters.

> - `POST /logs/stderr`: Update the log level used in the console. It takes one parameter, `level`, which is the log level you want to use: `trace,` `debug,` `info,` `warn,` `error`, or `off.`

Author: Not sure if it's only about `logs/stderr` or all the routes, but I guess it's only stderr based on our discussion on GitHub

🗣️ This feature is experimental, and we need your help to improve it! Share your thoughts and feedback on this [GitHub discussion](https://github.com/orgs/meilisearch/discussions/721).

⚠️ Experimental features may be incompatible between Meilisearch versions.

## Other improvements

* Related to the [Prometheus experimental feature](https://github.com/orgs/meilisearch/discussions/625): add job variable to Grafana dashboard (#4330) @capJavert
* Multiple language support improvements, including the Vietnamese Normalization (Ð and Đ into d). Now uses Charabia [v0.8.7](https://github.com/meilisearch/charabia/releases/tag/v0.8.7). (#4365) @agourlay, @choznerol, @ngdbao, @timvisee, @xshadowlegendx, and @ManyTheFish

# Fixes 🐞

TBD

# Misc

* Dependencies upgrade
  * Bump rustls-webpki from 0.101.3 to 0.101.7 (#4263)
  * Bump h2 from 0.3.20 to 0.3.24 (#4345)
  * Update the dependencies (#4332) @Kerollmops
* CIs and tests
  * Update SDK test dependencies (#4293) @curquiza
* Documentation
  * Add Setting API reminder in issue template (#4325) @ManyTheFish

>   * Update README (#4319) @codesmith-emmy

Author: ```suggestion
  * Update README (#4319) @codesmith-emmy
  *  Fix broken links in README.md (#4360) @Elliot67
```

Meilisearch v1.7.0 mostly focuses on improving v1.6.0 features, indexing speed and hybrid search. GPU computing is now supported.

<!-- The following lines should NOT be put in the PRE-release changelogs -->
🧰 All official Meilisearch integrations (including SDKs, clients, and other tools) are compatible with this Meilisearch release. Integration deployment happens between 4 to 48 hours after a new version becomes available.

<!-- The following lines should NOT be put in the PRE-release changelogs -->
Some SDKs might not include all new features—consult the project repository for detailed information. Is a feature you need missing from your chosen SDK? Create an issue letting us know you need it, or, for open-source karma points, open a PR implementing it (we'll love you for that ❤️).

# New features and improvements 🔥

## Improve AI with Meilisearch (experimental feature)

> 🗣️ AI work is still experimental and we need your help to improve it! Share your thoughts and feedback on this [GitHub discussion](https://github.com/orgs/meilisearch/discussions/677).
> 
> To use it, you need to enable `vectorSearch` through [the `/experimental-features` route](https://www.meilisearch.com/docs/reference/api/experimental_features).
> 
> 💡 More documentation about AI search with Meilisearch [here](https://www.meilisearch.com/docs/learn/experimental/vector_search#vector-search-with-auto-embeddings).

Author: ```suggestion
To activate AI-powered search, set `vectorStore` to `true` in [the `/experimental-features` route](https://www.meilisearch.com/docs/reference/api/experimental_features). Consult the [Meilisearch documentation](https://www.meilisearch.com/docs/learn/experimental/vector_search#vector-search-with-auto-embeddings) for more information.

🗣️ This is an experimental feature and we need your help to improve it! Share your thoughts and feedback on this [GitHub discussion](https://github.com/orgs/meilisearch/discussions/677).
```

- `text-embedding-3-large` with a default dimension of 3072.

The new models:
- are cheaper
- produce more relevant results in standardized tests
- allow to set up the dimensions of the embeddings to control the trade-off between accuracy and performance (including storage)

It means that it is now possible to pass the dimensions field when using the `OpenAi` source. This was previously only available for the `userProvided` source in the previous releases.

There are some rules, though, we detail with these examples:

```json5
"embedders": {
  "large": {
    "source": "openAi",
    "model": "text-embedding-3-large",
    "dimensions": 512 // must be >0, must be <= 3072 for "text-embedding-3-large"
  },
  "small": {
    "source": "openAi",
    "model": "text-embedding-3-small",
    "dimensions": 1024 // must be >0, must be <= 1536 for "text-embedding-3-small"
  },
  "legacy": {
    "source": "openAi",
    "model": "text-embedding-ada-002",
    "dimensions": 1536 // must =1536  for "text-embedding-ada-002"
  },
  "omitted_dimensions": { // uses the default dimension
    "source": "openAi",
    "model": "text-embedding-ada-002",

>   }

Author: Not sure why we want to remove some of them. They are relevant to me

>   }

Reviewer 2: Usually it's best examples are as short as possible, as this helps with readability. I think this is particularly true in changelogs, which are not supposed to be sources of exhaustive info. I'm also unsure users will benefit from outlining several permutations of the same pair of settings (`model`, `dimension`).

I'll leave the decision to you, though.

}

> ```

Author: @guimachiavelli why do you say this? Did you see something in the PRD? Because when I see the examples I wrote (I copied/pasted them), you can set up dimension for `text-embedding-ada-002`

> ```

Reviewer 2: This is for the sake of simplicity and giving users unambiguous instructions. Technically, you may set `dimensions` when using `ada-002`, but it must be exactly the model's size. Meaning, yes, you may use it, but it's not going to do anything.

We're also covering this in the docs: https://github.com/meilisearch/documentation/pull/2744#discussion_r1511475771

> ```

Author: From my POV, when I read "You cannot", it means you will not be able to do it at all. And it's false, you can. That's why I'm confused

> ```

Author: Hum, this is where I had a miss-understanding -> I thought we could put something different in `dimension`, but we can only put `1536`, I just re-read correctly [the issue](https://github.com/meilisearch/meilisearch/issues/4394)

So I agree with you!

**Steps**

1. Follow [the guide](https://huggingface.github.io/candle/guide/installation.html) to install the CUDA dependencies
2. Clone Meilisearch: `git clone https://github.com/meilisearch/meilisearch.git`
3. Compile Meilisearch with the `cuda` feature: `cargo build --release --package meilisearch --features cuda`
4. In the freshly compiled Meilisearch, enable the vector store experimental feature:
```bash
❯ curl \
  -X PATCH 'http://localhost:7700/experimental-features/' \
  -H 'Content-Type: application/json'  \
--data-binary '{ "vectorStore": true }'
```
5. Add an HuggingFace embedder to the settings:
```bash
curl \
-X PATCH 'http://localhost:7700/indexes/your_index/settings/embedders' \
-H 'Content-Type: application/json' --data-binary \
'{ "default": { "source": "huggingFace" } }'
```

Done by @dureuill in #4304.

## Improve indexing speed & reduce memory crashes

* Auto-batch the task deletions to reduce indexing time (#4316) @irevoire
* Improve indexing speed for vector store (makes the [Hybrid search experimental feature](https://github.com/orgs/meilisearch/discussions/677) indexing time more than 10 times faster) (#4332) @Kerollmops @irevoire

> * Reduce memory usage, so reduce the memory crashes, by caping the maximum memory of the grenad sorters (#4388) @Kerollmops

Author: I would rather we explain what it means for users, so the reduction of memory usage

> * Reduce memory usage, so reduce the memory crashes, by caping the maximum memory of the grenad sorters (#4388) @Kerollmops

Reviewer 2: No problem. I updated the suggestion to explicitly state that.

View detailed scores per ranking rule for each document with the `showRankingScoreDetails` search parameter:

```bash
curl \
  -X POST 'http://localhost:7700/indexes/movies/search' \
  -H 'Content-Type: application/json' \
  --data-binary '{ "q": "Batman Returns", "showRankingScoreDetails": true }'
```

When `showRankingScoreDetails` is set to `true`, returned documents include a `_rankingScoreDetails` field. This field contains score values for each ranking rule.

```json
"_rankingScoreDetails": {
  "words": {
    "order": 0,
    "matchingWords": 1,
    "maxMatchingWords": 1,
    "score": 1.0
  },
  "typo": {
    "order": 1,
    "typoCount": 0,
    "maxTypoCount": 1,
    "score": 1.0
  },
  "proximity": {
    "order": 2,
    "score": 1.0
  },
  "attribute": {
    "order": 3,
    "attributes_ranking_order": 0.8,
    "attributes_query_word_order": 0.6363636363636364,
    "score": 0.7272727272727273
  },
  "exactness": {
    "order": 4,
    "matchType": "noExactMatch",
    "matchingWords": 0,
    "maxMatchingWords": 1,
    "score": 0.3333333333333333
  }
}
```

Done by @dureuill in #4389.

## Logs improvements

> We made some changes regarding our logs to help with debugging and bug reporting.
> 
> Done by @irevoire in #4391
> 
> ### Log format change
> 
> ⚠️ If you did any automation based on Meilisearch logs, be aware of the changes. More information [here](https://meilisearch.notion.site/Logs-trace-changes-in-v1-7-0-6dd111ecd66448aba30d2f3cef212749).
> 
> The default log format evolved slightly from this:
> 
> ```bash
> [2024-02-06T14:54:11Z INFO  actix_server::builder] starting 10 workers
> ```
> 
> To this:
> 
> ```bash
> 2024-02-06T13:58:14.710803Z  INFO actix_server::builder: 200: starting 10 workers
> ```

Author: ````suggestion
Done by @irevoire in #4391

### Log output modified

Log messages now follow a different pattern:

```bash
# new format ✅ 
2024-02-06T14:54:11Z INFO actix_server::builder: 200: starting 10 workers
# old format ❌ 
[2024-02-06T14:54:11Z INFO  actix_server::builder] starting 10 workers
```

⚠️ This change may impact you if you have any automated tasks based on log output.
````

### Experimental: choose your log format

And one new CLI parameter is introduced called `--experimental_logs_format`. This lets you specify the log format you want to use on the console.

Accepted values are:

- `human`: the default human-readable one
- `json`: JSON structured logs

🗣️ This feature is experimental, and we need your help to improve it! Share your thoughts and feedback on this [GitHub discussion](https://github.com/orgs/meilisearch/discussions/723).

⚠️ Experimental features may be incompatible between Meilisearch versions.

### Experimental: new routes to manage logs

> This new version of Meilisearch introduces 3 new experimental routes
> - `POST /logs/stream`: streams the log happening in real-time. Requires two parameters:
>   - `target`: defines the log level and on which part of the engine you want to apply it. It takes the form of `code_part=log_level`, for example, `index_scheduler=info`. If you want to apply the changes to the whole engine, omit `code_part=`. Valid values for log level are: `trace,` `debug,` `info,` `warn,` `error`, or `off`.
>   - `mode`: selects in what format of log you want. Two options are available: `fmt` (basic logs) or `profile` (a way more complex trace)
> - `DELETE /logs/stream`: stops the listener from the meilisearch perspective. Does not require any parameters.
> - `POST /logs/stderr`: Update the log level used in the console. It takes one parameter, `target`, which is the same as the target of the `POST /logs/stream` route.
> 
> An example to disable any log from actix, but to keep all other logs at the `DEBUG` level:
> 
> ```bash
> curl \
>   -X POST http://localhost:7700/logs/stream \
>   -H 'Content-Type: application/json' \
>   --data-binary '{
>       "mode": "human",
>       "target": "actix=off,debug"
>     }'
> ```
> 
> 💡 *More information in the [New experimental routes section of this file](https://meilisearch.notion.site/Logs-trace-changes-in-v1-7-0-6dd111ecd66448aba30d2f3cef212749).*
> 
> ⚠️ Some remarks on this `POST /logs/stream` route:
> - You can have only one listener at a time
> - Listening to the route doesn’t seem to work with `xh` or `httpie` for the moment
> - When killing the listener, it may stay installed on Meilisearch for some time, and you will need to call the `DELETE /logs/stream` route to get rid of it.
> 
> 🗣️ This feature is experimental, and we need your help to improve it! Share your thoughts and feedback on this [GitHub discussion](https://github.com/orgs/meilisearch/discussions/721).
> 
> ⚠️ Experimental features may be incompatible between Meilisearch versions.

Reviewer 2: ````suggestion
Meilisearch v1.7 introduces 2 new experimental API routes: `/logs/stream` and `/logs/stderr`. 

Use the `/experimental-features` route to activate both routes during runtime:

```bash
curl \
  -X PATCH 'http://localhost:7700/experimental-features/' \
  -H 'Content-Type: application/json'  \
--data-binary '{
    "logsRoute": true
  }'
```

🗣️ This feature is experimental, and we need your help to improve it! Share your thoughts and feedback on this [GitHub discussion](https://github.com/orgs/meilisearch/discussions/721).

⚠️ Experimental features may be incompatible between Meilisearch versions.

#### `/logs/stream`

Use the `POST` endpoint to output logs in a stream. The following example disables actix logging and keeps all other logs at the `DEBUG` level:

```bash
curl \
  -X POST http://localhost:7700/logs/stream \
  -H 'Content-Type: application/json' \
  --data-binary '{
      "mode": "human",
      "target": "actix=off,debug"
    }'
```

This endpoint requires two paramaters:
- `target`: defines the log level and on which part of the engine you want to apply it. Must be a string formatted as `code_part=log_level`. Omit `code_part=` to set a single log level for the whole strram. Valid values for log level are: `trace,` `debug,` `info,` `warn,` `error`, or `off`
- `mode`: accepts `fmt` (basic) or `profile` (verbose trace)

Use the `DELETE` endpoint of `/logs/stream` to interrupt a stream:
```bash
curl -X DELETE http://localhost:7700/logs/stream
```

You may only have one listener at a time. Meilisearch log streams are not compatible with `xh` or `httpie`.

#### `/logs/stderr`

Use the `POST` endpoint to configure the default log output for non-stream logs:

```bash
curl \
  -X POST http://localhost:7700/logs/stream \
  -H 'Content-Type: application/json' \
  --data-binary '{
      "target": "debug"
    }'
```

`/logs/stderr` accepts one parameter:
- `target`: defines the log level and on which part of the engine you want to apply it. Must be a string formatted as `code_part=log_level`. Omit `code_part=` to set a single log level for the whole strram. Valid values for log level are: `trace,` `debug,` `info,` `warn,` `error`, or `off`
````

## Other improvements

> * Related to the [Prometheus experimental feature](https://github.com/orgs/meilisearch/discussions/625): add job variable to Grafana dashboard (#4330) @capJavert

Reviewer 2: ```suggestion
* [Prometheus](https://github.com/orgs/meilisearch/discussions/625): add job variable to Grafana dashboard (#4330) @capJavert
```

Author: I would rather have the reference to experimental feature

Author: ```suggestion
* [Prometheus experimental feature](https://github.com/orgs/meilisearch/discussions/625): add job variable to Grafana dashboard (#4330) @capJavert
```

> * Multiple language support improvements, including the Vietnamese Normalization (Ð and Đ into d). Now uses Charabia [v0.8.7](https://github.com/meilisearch/charabia/releases/tag/v0.8.7). (#4365) @agourlay, @choznerol, @ngdbao, @timvisee, @xshadowlegendx, and @ManyTheFish

Reviewer 2: ```suggestion
* Multiple language support improvements, including expanded Vietnamese normalization (Ð and Đ into d). Now uses Charabia [v0.8.7](https://github.com/meilisearch/charabia/releases/tag/v0.8.7). (#4365) @agourlay, @choznerol, @ngdbao, @timvisee, @xshadowlegendx, and @ManyTheFish
```

# Fixes 🐞

TBD

# Misc

* Dependencies upgrade
  * Bump rustls-webpki from 0.101.3 to 0.101.7 (#4263)
  * Bump h2 from 0.3.20 to 0.3.24 (#4345)
  * Update the dependencies (#4332) @Kerollmops
* CIs and tests
  * Update SDK test dependencies (#4293) @curquiza
* Documentation
  * Add Setting API reminder in issue template (#4325) @ManyTheFish
  * Update README (#4319) @codesmith-emmy
  * Fix broken links in README.md (#4360) @Elliot67
* Misc
  * Fix compilation warnings (#4295) @irevoire

❤️ Thanks again to our external contributors:

> - [Meilisearch](https://github.com/meilisearch/meilisearch): @capJavert, @codesmith-emmy and @Gosti

Author: ```suggestion
- [Meilisearch](https://github.com/meilisearch/meilisearch): @capJavert, @codesmith-emmy, @Elliot67 and @Gosti
```

  },
  "legacy": {
    "source": "openAi",
    "model": "text-embedding-ada-002",
    "dimensions": 1536 // must =1536  for "text-embedding-ada-002"
  },
  "omitted_dimensions": { // uses the default dimension
    "source": "openAi",
    "model": "text-embedding-ada-002",
  }
}
```

Done in #4375 by @Gosti.

### GPU support when computing Hugging Face embeddings

Activate CUDA to use Nvidia GPUs when computing Hugging Face embeddings. This can significantly improve embedding generation speeds.

To enable GPU support through CUDA for HuggingFace embedding generation:

1. [Install CUDA dependencies](https://huggingface.github.io/candle/guide/installation.html)
2. Clone and compile Meilisearch with the `cuda` feature: `cargo build --release --package meilisearch --features cuda`
3. Launch your freshly compiled Meilisearch binary
4. [Activate vector search](https://www.meilisearch.com/docs/learn/experimental/vector_search#activate-vector-search)
5. [Add a Hugging Face embedder](https://www.meilisearch.com/docs/learn/experimental/vector_search#generate-auto-embeddings-with-huggingface)

Done by @dureuill in #4304.

> 

Reviewer 3: Additions related to [v1.7.0rc.2](https://github.com/meilisearch/meilisearch/releases/tag/v1.7.0-rc.2):
```suggestion

### Additional analytics related to the GPU support
* [Add a boolean analytics `gpu_enabled` when Meilisearch is compiled with CUDA](https://github.com/meilisearch/meilisearch/pull/4443) by @dureuill
```

Author: Nope, we don't add this kind of information to the changelogs. Not really user facing. The docs however should be updated.

Reviewer 3: Yes, I was wondering about this one, I put it just in case

Author: And you were right, it's better to add it than forget it!
